alvr/audio/src/linux.rs

```diff
@@ -0,0 +1,350 @@
+            pw_stream_state,
+            sample_rate,
+            channels_count,
+            pw_receiver,
+            pw_loop_buffer_arc,
+        ) {
+            Ok(_) => {
+                debug!("Pipewire loop exiting");
+            }
+            Err(e) => error!("Pipewire error: {}", e.to_string()),
+        }
+    });
+
+    while running() {
+        let stream_audio = {
+            || {
+                if let Some(stream_state) = pw_stream_state_arc.try_lock() {
+                    *stream_state == StreamState::Streaming && running()
+                } else {
+                    false
+                }
+            }
+        };
+        let receive_samples_buffer_arc = Arc::clone(&receive_samples_buffer_arc);
+        receive_samples_loop(
+            stream_audio,
+            receiver,
+            receive_samples_buffer_arc,
+            channels_count as _,
+            batch_frames_count,
+            average_buffer_frames_count,
+        )
+        .ok();
+    }
+
+    if let Err(_) = pw_sender.send(Terminate) {
+        error!(
+            "Couldn't send pipewire termination signal, deinitializing forcefully.
+            Restart VR app to re-initialize pipewire."
+        );
+        unsafe { pw::deinit() };
+    }
+
+    match thread.join() {
+        Ok(_) => debug!("Pipewire thread joined"),
+        Err(_) => {
+            error!("Couldn't wait for pipewire thread to finish");
+        }
+    }
+    Ok(())
+}
+
+fn pw_microphone_loop(
+    pw_stream_state: Arc<Mutex<StreamState>>,
+    sample_rate: u32,
+    channels_count: u16,
+    pw_receiver: pw::channel::Receiver<Terminate>,
+    sample_buffer: Arc<Mutex<VecDeque<f32>>>,
+) -> Result<(), pw::Error> {
+    debug!("Staring microphone pw-thread");
+    let mainloop = pw::main_loop::MainLoop::new(None)?;
+
+    let _receiver = pw_receiver.attach(mainloop.as_ref(), {
+        let mainloop = mainloop.clone();
+        move |_| mainloop.quit()
+    });
+
+    let context = pw::context::Context::new(&mainloop)?;
+    let core = context.connect(None)?;
+
+    let stream = pw::stream::Stream::new(
+        &core,
+        "alvr-mic",
+        pw::properties::properties! {
+            *pw::keys::NODE_NAME => "ALVR Microphone",
+            *pw::keys::MEDIA_NAME => "alvr-mic",
+            *pw::keys::MEDIA_TYPE => "Audio",
+            *pw::keys::MEDIA_CATEGORY => "Playback",
+            *pw::keys::MEDIA_CLASS => "Audio/Source",
+            *pw::keys::MEDIA_ROLE => "Communication",
+        },
+    )?;
+
+    let chan_size = std::mem::size_of::<f32>();
+    let default_channels_count: usize = channels_count.into();
+    // Amount of bytes one full processing will take
+    let stride = chan_size * default_channels_count;
+    let _listener: StreamListener<f32> = stream
+        .add_local_listener()
+        .state_changed(move |_, _, old_state, new_state| {
+            debug!("state changed, old: {:?}, new: {:?}", old_state, new_state);
```

Comment:
This logging is a bet messy

Reply:
removed most of the logging, most of it was only needed during initial prototyping

Reply:
This particular instance needs to go tho or have a notice that it's about pipewire